File: tasks-api-java/src/main/java/br/com/darlan/tasks/controller/TaskController.java
package br.com.darlan.tasks.controller;


import br.com.darlan.tasks.aspect.annotations.LogExecution;
import br.com.darlan.tasks.dto.TaskDTO;
import br.com.darlan.tasks.service.TaskService;
import br.com.darlan.tasks.util.TaskUtil;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.List;

@RestController
@RequestMapping("/tasks")
public class TaskController {
    private final TaskService service;
    private final TaskUtil taskUtil;

    @Autowired
    public TaskController(final TaskService service, final TaskUtil taskUtil) {
        this.service = service;
        this.taskUtil = taskUtil;
    }

    @GetMapping
    @LogExecution
    public ResponseEntity<List<TaskDTO>> list() {
        List<TaskDTO> dtos = taskUtil.entityListToDTOList(service.list());
        dtos = taskUtil.addSelfLink(dtos);
        return ResponseEntity.ok(dtos);
    }

    @GetMapping("/{idTask}")
    @LogExecution
    public ResponseEntity<TaskDTO> findById(@PathVariable("idTask") Long idTask) {
        return ResponseEntity.ok(taskUtil.entityToDTO(service.findById(idTask)));
    }

    @PostMapping
    @LogExecution
    public ResponseEntity<TaskDTO> save(@RequestBody TaskDTO dto) {
        return ResponseEntity.ok(taskUtil.entityToDTO(service.save(taskUtil.dtoToEntity(dto))));
    }

    @PutMapping("/{idTask}")
    @LogExecution
    public ResponseEntity<TaskDTO> update(@PathVariable("idTask") Long idTask, @RequestBody TaskDTO dto) {
        return ResponseEntity.ok(taskUtil.entityToDTO(service.update(idTask, taskUtil.dtoToEntity(dto))));
    }

    @DeleteMapping("/{idTask}")
    @LogExecution
    public ResponseEntity<String> delete(@PathVariable("idTask") Long idTask) {
        service.delete(idTask);
        return ResponseEntity.ok("Task successfully deleted");
    }
}

File: kotlin/tasks-api-kotlin/src/main/kotlin/br/com/darlan/tasks/controller/TaskController.kt
package br.com.darlan.tasks.controller

import br.com.darlan.tasks.aspect.annotations.LogExecution
import br.com.darlan.tasks.dto.TaskDTO
import br.com.darlan.tasks.service.TaskService

import org.springframework.beans.factory.annotation.Autowired
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.DeleteMapping
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.PutMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController

@RestController
@RequestMapping("/tasks")
class TaskController (@Autowired var service: TaskService){

    @GetMapping
    @LogExecution
    fun list(): ResponseEntity<List<TaskDTO>> {
        return ResponseEntity.ok(service.list())
    }

    @PostMapping
    @LogExecution
    fun save(@RequestBody dto: TaskDTO): ResponseEntity<TaskDTO> {
        return ResponseEntity.ok(service.save(dto))
    }

    @GetMapping("/{idTask}")
    @LogExecution
    fun findById(@PathVariable("idTask") idTask: Long): ResponseEntity<TaskDTO> {
        return ResponseEntity.ok(service.findById(idTask))
    }

    @PutMapping("/{idTask}")
    @LogExecution
    fun update(@PathVariable("idTask") idTask: Long, @RequestBody dto: TaskDTO): ResponseEntity<TaskDTO> {
        return ResponseEntity.ok(service.update(idTask, dto))
    }

    @DeleteMapping("/{idTask}")
    @LogExecution
    fun delete(@PathVariable("idTask") idTask: Long): ResponseEntity<String> {
        service.delete(idTask)
        return ResponseEntity.ok("Task $idTask deleted")
    }
}